# Coding Exercises (Part 1)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 5: Conditional Statements & while loops

The ABC Company has a mature project with declining revenues and increasing costs. Net Cashflows are projected as follows:

- 500 in year 1,
- 350 in year 2, 
- 200 in year 3,
- 50 in year 4,
- -100 in year 5,
- -200 in year 6

1. Separate __positive__ and __negative__ cashflows. __Create two new lists__!

In [3]:
# Run the cell
cf = [500, 350, 200, 50, -100, -200]

[500, 350, 200, 50, -100, -200]

In [22]:
def foldl(f, acc, xs):
  if not xs:
    return acc
  else:
    return foldl(f, f(acc, xs[0]), xs[1 :])

negatives, positives = foldl(lambda acc, x: (acc[0] + [x], acc[1]) if x < 0 else (acc[0], acc[1] + [x]), ([], []), cf)

print(negatives)
print(positives)

# Asked to "partition" by mutation, but we'll do it a FP way (as above).
cf_pos =[]
cf_neg =[]

# for i in cf: # complete the code!

[-100, -200]
[500, 350, 200, 50]


In [69]:
from itertools import dropwhile

# firstUnprofitableYear = (list(dropwhile(lambda xs: xs[1] > 0, list(enumerate(cf)))) or [None])[0]
firstUnprofitableYear = next(dropwhile(lambda xs: xs[1] > 0, list(enumerate(cf, 1))), None)
print(firstUnprofitableYear)
f"Year {firstUnprofitableYear[0]} the first unprofitable year" if firstUnprofitableYear is not None else None


(5, -100)


'Year 5 the first unprofitable year'

2. Get the __first unprofitable year__. Your Code should return the statement __"Year ... is the first unprofitable Year!"__ 

In [68]:
from itertools import takewhile

lastUnprofitableYear = next(reversed(list(takewhile(lambda x: x[1] > 0, list(enumerate(cf, 1))))), None)
print(lastUnprofitableYear)
f"Year {lastUnprofitableYear[0]} is the last profitable year" if lastUnprofitableYear is not None else None

(4, 50)


'Year 4 is the last profitable year'

3. Get the __last profitable year__. Your Code should return the statement __"Year ... is the last profitable Year!"__ 

The __modulo operator %__ returns the __remainder__ of the division between two numbers. For instance, 5 % 3 returns 2. 

In [70]:
# run the cell
5 % 3

2

In [71]:
# run the cell
14 % 5

4

If a is divisible by b __without remainder__, a % b __returns 0__. 

In [72]:
# run the cell
a = 9
b = 3
a % b

0

4. Return the __first number n__ that is __divisible by 362 and 216__ (without remainder).<br>
Your code shall return the following statement: __"... is the first number divisible by 362 and 216!"__

In [1]:
from tail_recursive import tail_recursive

a = 362
b = 216

n = 1

# while True: # Complete the code! Actually, I want to avoid using "while"

@tail_recursive
def go(n) -> int:
  if n % a == 0 and n % b == 0:
    return n
  else:
    return go.tail_call(n + 1)

print(go(n))
    

39096


(result: 39096)

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

The ABC Company has a mature project with declining revenues and increasing costs. Net Cashflows are projected as follows:

- 500 in year 1,
- 350 in year 2, 
- 200 in year 3,
- 50 in year 4,
- -100 in year 5,
- -200 in year 6

1. Separate __positive__ and __negative__ cashflows. __Create two new lists__!

In [52]:
#run the cell
cf = [500, 350, 200, 50, -100, -200]

In [53]:
cf_pos =[]
cf_neg =[]

for i in cf:
    if i > 0:
        cf_pos.append(i)
    elif i < 0:
        cf_neg.append(i)
print(cf_pos, cf_neg)

[500, 350, 200, 50] [-100, -200]


2. Get the __first unprofitable year__. Your Code should return the statement __"Year ... is the first unprofitable Year!"__ 

In [54]:
n = 1
for i in cf:
    if i < 0:
        print("Year {} is the first unprofitable year!".format(n))
        break
    n += 1

Year 5 is the first unprofitable year!


3. Get the __last profitable year__. Your Code should return the statement __"Year ... is the last profitable Year!"__ 

In [67]:
n = 1
for i in cf:
    if i < 0:
        print("Year {} is the last profitable year!".format(n - 1))
        break
    n += 1

Year 4 is the last profitable year!


The __modulo operator %__ returns the __remainder__ of the division between two numbers. For instance, 5 % 3 returns 2. 

In [ ]:
# run the cell
5 % 3

In [ ]:
# run the cell
14 % 5

If a is divisible by b __without remainder__, a % b __returns 0__. 

In [ ]:
# run the cell
a = 9
b = 3
a % b

4. Return the __first number n__ that is __divisible by 362 and 216__ (without remainder).<br>
Your code shall return the following statement: __"... is the first number divisible by 362 and 216!"__

In [ ]:
a = 362
b = 216

n = 1
while True:
    if (n%a == 0) and (n%b == 0):
        print("{} is the first number divisible by {} and {}!".format(n, a, b))
        break
    n+=1

(result: 39096)